src/cbor/decoder.rs

```diff
@@ -24,105 +24,113 @@ impl DecoderCursor {
+            let new_item = self.decode_item()?;
+            array.push(new_item);
         }
         Ok(CborType::Array(array))
     }
 
-    /// Read a byte string and return it as hex string.
-    fn read_byte_string(&mut self) -> Result<CborType, &'static str> {
-        let length = self.read_int().unwrap();
+    /// Read a byte string and return it.
+    fn read_byte_string(&mut self) -> Result<CborType, CborError> {
+        let length = self.read_int()?;
         if length > usize::max_value() as u64 {
-            return Err("Byte array is too large to allocate.");
+            return Err(CborError::InputTooLarge);
         }
         let length = length as usize;
-        let mut byte_string: Vec<u8> = Vec::with_capacity(length);
-        // XXX: rewrite without unsafe.
-        unsafe {
-            byte_string.set_len(length);
-        }
-        if self.cursor.read(&mut byte_string).unwrap() != length {
-            return Err("Couldn't read enough data for this byte string");
-        }
+        let mut byte_string: Vec<u8> = vec![0; length];
+        match self.cursor.read_exact(&mut byte_string) {
+            Err(_) => return Err(CborError::TruncatedInput),
+            Ok(()) => {}
+        };
         Ok(CborType::Bytes(byte_string))
     }
 
     /// Read a map.
-    fn read_map(&mut self) -> Result<CborType, &'static str> {
-        // XXX: check for duplicate keys.
-        let num_items = self.read_int().unwrap();
+    fn read_map(&mut self) -> Result<CborType, CborError> {
+        let num_items = self.read_int()?;
         // Create a new array.
         let mut map: BTreeMap<CborType, CborType> = BTreeMap::new();
         // Decode each of the num_items (key, data item) pairs.
         for _ in 0..num_items {
-            let key_val = self.decode_item().unwrap();
-            let item_value = self.decode_item().unwrap();
+            let key_val = self.decode_item()?;
+            let item_value = self.decode_item()?;
+            if map.contains_key(&key_val) {
+                return Err(CborError::DuplicateMapKey);
+            }
             map.insert(key_val, item_value);
         }
         Ok(CborType::Map(map))
     }
 
-    /// Decodes the next item in DecoderCursor.
-    pub fn decode_item(&mut self) -> Result<CborType, &'static str> {
-        let pos = self.cursor.position() as usize;
-        let major_type = self.cursor.get_ref()[pos] >> 5;
-        match major_type {
-            0 => return Ok(CborType::Integer(self.read_int().unwrap())),
-            1 => return Ok(self.read_signed_int().unwrap()),
-            2 => return Ok(self.read_byte_string().unwrap()),
-            4 => return Ok(self.read_array().unwrap()),
-            5 => return Ok(self.read_map().unwrap()),
+    /// Peeks at the next byte in the cursor, but does not change the position.
+    fn peek_byte(&mut self) -> Result<u8, CborError> {
+        let mut x: Vec<u8> = vec![0; 1];
+        match self.cursor.read_exact(&mut x) {
```

Comment:
We have the ` match cursor.read` pattern a couple times. Would it make sense to implement the `From` trait for the `Result` to match it to `CborError`?

Reply:
Good call. Rather than a `From` trait, I ended up just refactoring it more, so there's only one `self.cursor.read_exact` call (I was concerned about a generic `io::Result` -> `CborError` conversion being too broad, but we can revisit if it ends up being useful).